pkg/controller/deployment/util/deployment_util.go

```diff
@@ -261,14 +261,25 @@ func SetNewReplicaSetAnnotations(deployment *apps.Deployment, newRS *apps.Replic
 	// If a revision annotation already existed and this replica set was updated with a new revision
 	// then that means we are rolling back to this replica set. We need to preserve the old revisions
 	// for historical information.
-	if ok && annotationChanged {
+	if ok && oldRevisionInt < newRevisionInt {
 		revisionHistoryAnnotation := newRS.Annotations[RevisionHistoryAnnotation]
 		oldRevisions := strings.Split(revisionHistoryAnnotation, ",")
 		if len(oldRevisions[0]) == 0 {
 			newRS.Annotations[RevisionHistoryAnnotation] = oldRevision
 		} else {
-			oldRevisions = append(oldRevisions, oldRevision)
-			newRS.Annotations[RevisionHistoryAnnotation] = strings.Join(oldRevisions, ",")
+			totalLen := len(revisionHistoryAnnotation) + len(oldRevision) + 1
+			// index for the starting position in oldRevisions
+			start := 0
+			for totalLen > revHistoryLimit && start < len(oldRevisions) {
+				totalLen = totalLen - len(oldRevisions[start]) - 1
```

Comment:
If we cap this by a reasonable fixed number of elements, this code would be a lot simpler:
```go
oldRevisions = append(oldRevisions, oldRevision)
if len(oldRevisions) > revHistoryLimit {
    oldRevisions = oldRevisions[len(oldRevisions) - revHistoryLimit:]
}
```

Reply:
Since we don't know per element length, using element count is inaccurate.